Forwarded: https://github.com/BuddiesOfBudgie/budgie-desktop/pull/128
Author: David Mohammed <fossfreedom@ubuntu.com>
Last-Update: 2022-02-22
Description: [PATCH1/6] Support freedesktop color-scheme latest libhandy versions and
 libadwaita-1 use the new freedesktop spec to alter the color scheme of apps
 compiled with these libraries. This commit builds upon the existing
 dark-theme flag budgie uses to now indicate that apps (if supported) should
 prefer a dark styling of libhandy/libadwaita.

---
 src/theme/meson.build     |  8 ++++++++
 src/theme/theme-manager.c | 25 ++++++++++++++++++++++++-
 2 files changed, 32 insertions(+), 1 deletion(-)

--- a/src/theme/meson.build
+++ b/src/theme/meson.build
@@ -1,6 +1,14 @@
 gnome = import('gnome')
 pkgconfig = import('pkgconfig')
 
+config_h = configuration_data()
+config_h.set('GSD42', dep_gsd.version().version_compare('>= 42'))
+# Write config.h
+configure_file(
+    output: 'config.h',
+    configuration: config_h,
+)
+
 # Compile to CSS dynamically.
 sass = find_program('sass', required: false)
 sass_args = ['--no-source-map', '--style', 'compressed', '@INPUT@', '@OUTPUT@']
--- a/src/theme/theme-manager.c
+++ b/src/theme/theme-manager.c
@@ -11,6 +11,7 @@
 
 #define _GNU_SOURCE
 
+#include "config.h"
 #include "theme-manager.h"
 #include "theme.h"
 #include <gtk/gtk.h>
@@ -23,12 +24,16 @@
 	GObject parent;
 	GtkCssProvider* css_provider;
 	GSettings* desktop_settings;
+	GSettings* ui_settings;
 	gboolean builtin_enabled;
 };
 
 static void budgie_theme_manager_set_theme_css(BudgieThemeManager* self, const gchar* theme_portion);
 static void budgie_theme_manager_theme_changed(BudgieThemeManager* self, GParamSpec* prop, GtkSettings* settings);
 static void budgie_theme_manager_builtin_changed(BudgieThemeManager* self, const gchar* key, GSettings* settings);
+#ifdef GSD42
+static void budgie_theme_manager_dark_theme_changed(BudgieThemeManager* self, GParamSpec* prop, GSettings* settings);
+#endif
 
 G_DEFINE_TYPE(BudgieThemeManager, budgie_theme_manager, G_TYPE_OBJECT)
 
@@ -74,6 +79,7 @@
 	/* TODO: Stop using .budgie-panel for desktop-wide schema ! */
 	self->desktop_settings = g_settings_new("com.solus-project.budgie-panel");
 	self->builtin_enabled = g_settings_get_boolean(self->desktop_settings, "builtin-theme");
+	self->ui_settings = g_settings_new("org.gnome.desktop.interface");
 
 	/* Update whether we can use the builtin theme or not */
 	g_signal_connect_swapped(self->desktop_settings, "changed::builtin-theme", G_CALLBACK(budgie_theme_manager_builtin_changed), self);
@@ -83,8 +89,11 @@
 
 	/* Bind the dark-theme option for the whole process */
 	g_settings_bind(self->desktop_settings, "dark-theme", settings, "gtk-application-prefer-dark-theme", G_SETTINGS_BIND_GET);
+#ifdef GSD42
+	g_signal_connect_swapped(self->desktop_settings, "changed::dark-theme", G_CALLBACK(budgie_theme_manager_dark_theme_changed), self);
 
-	/* Trigger theme changed */
+	budgie_theme_manager_dark_theme_changed(self, NULL, self->desktop_settings);
+#endif
 	budgie_theme_manager_theme_changed(self, NULL, settings);
 }
 
@@ -167,3 +176,17 @@
 	/* Update now based on whether we can use the built-in */
 	budgie_theme_manager_theme_changed(self, NULL, gtk_settings_get_default());
 }
+
+#ifdef GSD42
+static void budgie_theme_manager_dark_theme_changed(BudgieThemeManager* self, __attribute__((unused)) GParamSpec* prop, GSettings* settings) {
+	gboolean dark_theme = g_settings_get_boolean(settings, "dark-theme");
+
+	if (dark_theme) {
+		g_settings_set_string(self->ui_settings, "color-scheme", "prefer-dark");
+	}
+	else {
+		g_settings_reset(self->ui_settings, "color-scheme");
+	}
+
+}
+#endif
